**************************************************
Welcome to Applied OMICs Tutorial's documentation!
**************************************************

Welcome to the introductory tutorial for the curricular unit of Applied OMICs where you will learn fundamental bioinformatics analysis using mainly Linux command-line. During this tutorial, we will use real next-generation sequencing (NGS) retrieved from the Sequence Read Archive (SRA). In the end, you will able to assemble and analyze bacterial genomes.

If you have any doubts about this tutorial, you can contact me through my email **joana.mourao@uc.pt**


Prerequisites
#############

Almost none. This course is an introduction, so the only thing that you need to know is where to find your Terminal, the program that we will use to interact with the Shell (command-line interface).
If you are using a Unix-based operating system, it will be pretty straightforward to find it.
Contrarily if you are using Windows-based, you will need to look for a Windows Subsystem for Linux (WSL) or install for example Git Bash.


.. only:: builder_html

   Contents
   --------

.. toctree::
   :maxdepth: 2

   objectives

.. toctree::
   :maxdepth: 2
   :caption: Begginers

   before-begin

.. toctree::
   :numbered:
   :maxdepth: 2
   :caption: Advanced

   ngs-data
   ngs-tools
   ngs-qc
   ngs-taxonomy
   ngs-assembly
   ngs-annotations
   ngs-features
   ngs-snpvariants
   ngs-phylogeny
   ngs-pangenome

.. toctree::
   :maxdepth: 2
   :caption: Resources

   downloads

.. only:: latex

   .. raw:: latex

      \listoffigures
      \listoftables
